### DISCUSSION TITLE
How to modify a dotfile not through chezmoi?

### DISCUSSION THREAD

**Discussion by i2** [twpayne/chezmoi#1598] [Q&A]
[ Just to point out: I have read and am aware of [https://github.com/twpayne/chezmoi/issues/748#issuecomment-628065878](url) ]

Is it possible to modify a dotfile and be in sync with my dotfile repo, without using "chezmoi edit"? Please let me explain by the following example:

1. I have a blank dotfile repo on github and initialize chezmoi with it: `chezmoi init my_repo`
2. I add my tmux config to this repo: `chezmoi add ~/.tmux.conf`
3. all great so far. After a while I want to modify my tmux.conf
4. Habitually, I open my tmux.conf file and make my changes
5. At this point. if I run `chezmoi diff`, it sees that there are some changes between my git repo and local files.
6. `chezmoi update` will restore the local changes to the one from step 2

Is there a way to do this the other way around?
Git repo --- to --> local file is possible
local file --- to --> git repo without chezmoi edit ?

It doesn't feel natural to me to make note of what configs I manage with chezmoi and if I want to make some changes for those config files, make them using `chezmoi edit` instead of just vim/nano.

I have also looked at the following links before posting this:
- [https://github.com/twpayne/chezmoi/discussions/1312](url)
- [https://github.com/twpayne/chezmoi/discussions/1510](url)
- [https://github.com/twpayne/chezmoi/pull/1208](url)

Thank you!

**Comment by felipecrs** [comment]
The usual process is to edit the file in your git repository first then `chezmoi apply`. This is what `chezmoi edit` automates.

However, the opposite way is also simple:

1. Edit the file (e.g. `~/.zshrc`)
2. `chezmoi add .zshrc`

Just like git!

**Reply by i2** [reply]
> The usual process is to edit the file in your git repository first then `chezmoi apply`.

So, modify the dotfile and then git add commit push to my repo? I don't know, for some reason, this workflow doesn't feel natural to me. <---- kudos to @twpayne for creating chezmoi. It's a great tool. I'm just asking for clarifications to help me decide which workflow suits me better (my choices are chezmoi or [dac](https://github.com/sitaramc/notes/blob/master/dac)). Much respect.

Is there a way to trigger an auto-commit to my git repo, when I edit my dotfile say with vim (without issuing `chezmoi edit`)?

**Reply by felipecrs** [reply]
> What does chezmoi do then?

chezmoi does a million things. For example, you can use templates to modify your dotfiles according to the machine that you are using seamlessly (when it's a personal or work machine, an ephemeral container like codespaces, or macOS vs Linux).

> Is there a way to trigger an auto-commit to my git repo, when I edit my dotfile say with vim (without issuing chezmoi edit)?

Yes! The chezmoi docs are very extensive: https://www.chezmoi.io/docs/how-to/#automatically-commit-and-push-changes-to-your-repo

**Reply by felipecrs** [reply]
Replying to https://github.com/twpayne/chezmoi/discussions/1598#discussioncomment-1603202:

> Is this what workflow what you mean? According to this, it doesn't seem safe to do so.

Yes, that's what I meant and it's indeed not safe. With chezmoi, you can use templates inside the dotfiles in the git repository. If you wrote a template in your git repository, and then re-added the file (without templates), your custom templates would be lost in the source repository.

For example, check this section of my `~/.profile`:

https://github.com/felipecrs/dotfiles/blob/ddb1198a99efa3e3e001e59c191f744aca204452/home/dot_profile.tmpl#L3-L8

It will only be set when I'm on a WSL machine.

So, still, it's recommended to use `chezmoi edit`, or edit the files by yourself and run `chezmoi apply --verbose` after.

**Comment by i2** [comment]
Thanks for your response @felipecrs.

So again **re-add** the modified file? Like so:

1. (initially say on day 1 of using chezmoi) Add ~/.zshrc to git repo `chezmoi add ~/.zshrc`
2. After a while and some modifications to dofile again run `chezmoi add ~/.zshrc`

Is this what workflow what you mean? According to [this](https://github.com/twpayne/chezmoi/issues/748#issuecomment-628065878), it doesn't seem safe to do so.

**Reply by felipecrs** [reply]
Answered at https://github.com/twpayne/chezmoi/discussions/1598#discussioncomment-1603488.

**Comment by sisrfeng** [comment]
`chezmoi add ~/dotfile` does not update the modified files under chezmoi folder,  while `chezmoi re-add` doe.

Why is that ?

What is the difference  between `chezmoi add ~/dotfile` and `chezmoi re-add` ?  `chezmoi re-add` is convinient when we have to add lots of files ?

**Reply by twpayne** [reply]
> `chezmoi add ~/dotfile` does not update the modified files under chezmoi folder, while `chezmoi re-add` doe.
> 
> Why is that ?

A number of users complained when `chezmoi add` overwrote the changes they'd made to the source state. To make `chezmoi add` overwrite, use the `-f`/`--force` option.


> What is the difference between `chezmoi add ~/dotfile` and `chezmoi re-add` ? `chezmoi re-add` is convinient when we have to add lots of files ?

Exactly.

**Comment by xvzc** [comment]
In the latest release, `--watch` option for `chezmoi edit` command is available. it automatically applies the changes when the file has been saved.

If you use `neovim` or `vim` you can just define a function in your shell environment like below. 
I'm not very good at shell scripting by the way..
```bash
# "vi FILE_NAME" automatically runs "chemoi edit FILE_NAME" command if the file is under management by chemoi.
# otherwise just give the all arguments to neovim
function vi() {
  if [ $# -eq 0 ]; then
    nvim
    return
  fi

  # Added "--hardlink=false" option to get neovim undo available.
  chezmoi verify $1 > /dev/null 2>&1\
    && chezmoi edit --watch --hardlink=false $1\ 
    || nvim $@
}
```

**Reply by kwand** [reply]
Thanks, this was useful. In case anyone is using `fish`  like I am, here's a version converted to fish's scripting language (also not great at shell scripting, but it works):

```fish
function vi
    if test (count $argv) -eq 0 
        nvim
    else
        chezmoi verify $argv[1] &> /dev/null && chezmoi edit --watch --hardlink=false $argv[1] || nvim $argv
    end
end
```

**Reply by Lockszmith-GH** [reply]
Thanks! This is helpful,
I wrote too much script to deal with something like this, I mentioned it [here](https://github.com/twpayne/chezmoi/discussions/2631#discussioncomment-4722259)

**Comment by xvzc** [comment]
I wrote a neovim plugin that might help you edit chezmoi-managed files. Please try if anyone is interested.
https://github.com/xvzc/chezmoi.nvim

**Reply by twpayne** [reply]
Thanks, this is really nice. I've added it to [chezmoi's related software page](https://www.chezmoi.io/links/related-software/) (will be visible in about one hour).

**Reply by xvzc** [reply]
I appreciate that. It is a great honor for me!
